### PR TITLE
HIVE-25010: Create TestIcebergCliDriver and TestIcebergNegativeCliDriver

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
New iceberg specific TestIcebergCliDriver and TestIcebergNegativeCliDriver

### CODE REVIEW

File: itests/pom.xml [hunk 1]
```diff
@@ -248,6 +248,17 @@
         <version>${project.version}</version>
         <classifier>tests</classifier>
       </dependency>
+      <dependency>
```
Comment (pvary) [inline itests/pom.xml:251]: Wouldn't this dependency fail if the `hive-iceberg-handler` never compiled before on this machine?

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 2]
```diff
@@ -770,6 +770,8 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
         "If not set, defaults to the codec extension for text files (e.g. \".gz\"), or no extension otherwise."),
 
     HIVE_IN_TEST("hive.in.test", false, "internal usage only, true in test mode", true),
+    HIVE_IN_TEST_ICEBERG("hive.in.iceberg.test", false, "internal usage only, true when " +
```
Comment (pvary) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:773]: why is this needed?

File: data/conf/iceberg/hive-site.xml [hunk 3]
```diff
@@ -0,0 +1,321 @@
+<?xml version="1.0"?>
```
Comment (pvary) [inline data/conf/iceberg/hive-site.xml:1]: do we need our own hive-site at this stage? Can not we get away with a general one?

File: itests/pom.xml (hunk 1 above)
Comment (lcspinter) [reply itests/pom.xml:251]: I added the `iceberg` profile to the `.travis.yml`

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 2 above)
Comment (lcspinter) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:773]: During the creation of the iceberg table an exclusive lock is requested on the hms table. This logic is guarded by the value of `HIVE_IN_TEST` and `HINVE_IN_TEZ_TEST`. If any of these values are true and the operationtype is unset, we get an exception, which is the case in q tests. 
```java
f (lc.isSetOperationType() && lc.getOperationType() == DataOperationType.UNSET &&
                (MetastoreConf.getBoolVar(conf, ConfVars.HIVE_IN_TEST) || MetastoreConf.getBoolVar(conf, ConfVars.HIVE_IN_TEZ_TEST))) {
          throw new IllegalStateException("Bug: operationType=" + lc.getOperationType() + " for component "
                  + lc + " agentInfo=" + rqst.getAgentInfo());
        }
```
Changing the `HIVE_IN_TEST` to false is not the right solution, because it is evaluated in many other places as well, activating unwanted code paths.

File: data/conf/iceberg/hive-site.xml (hunk 3 above)
Comment (lcspinter) [reply data/conf/iceberg/hive-site.xml:1]: I needed a new conf, because of the `hive.in.iceberg.test=true` param.

File: itests/pom.xml (hunk 1 above)
Comment (pvary) [reply itests/pom.xml:251]: The Travis will be removed soon. That should not be a problem. My fear is that when the developers starts to run the the tests on their local machines without the `iceberg` flag, they will have a problem (maybe only if they are offline)

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 2 above)
Comment (pvary) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:773]: How I hate test code in the production codepath, I can not tell you...
@deniskuzZ: How hard would it be to get rid of this?

File: data/conf/iceberg/hive-site.xml (hunk 3 above)
Comment (pvary) [reply data/conf/iceberg/hive-site.xml:1]: I would rather get rid of the `hive.in.iceberg.test` altogether

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java (hunk 2 above)
Comment (marton-bod) [reply common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:773]: This caused trouble when migrating the iceberg-mr code to the new iceberg-handler module. The TestHiveShell configs had to patched to get around this as well..

File: iceberg/iceberg-handler/src/test/java/queries/negative/create_iceberg_table_failure.q [hunk 4]
```diff
@@ -0,0 +1,2 @@
+set hive.vectorized.execution.enabled=true;
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/queries/negative/create_iceberg_table_failure.q:1]: Can we add here other table creation failure scenarios too? One that comes to mind now is where you specify a PARTITIONED BY clause but also specify the json spec in the table properties. Another one could be related to type mismatches, i.e. using an unsupported Hive type, like interval in the columns

Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/queries/negative/create_iceberg_table_failure.q:1]: I suppose this was not meant to be comprehensive but just to set up the iceberg qtest infra? In that case, we can tackle these in upcoming PRs that beef up the test cases.

Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/queries/negative/create_iceberg_table_failure.q:1]: Yes, this PR provides the skeleton to run additional q tests from upcoming PRs.

File: data/conf/iceberg/hive-site.xml (hunk 3 above)
Comment (marton-bod) [reply data/conf/iceberg/hive-site.xml:1]: Agreed that @lcspinter will discuss with @deniskuzZ the effort involved in removing the problematic part from `TxnHandler`. Depending on the time estimate though, we might want to temporarily live with this workaround, because this patch is blocking several other patches of Laci.

**Review by pvary — APPROVED** [review]
+1 pending tests

File: data/conf/iceberg/hive-site.xml (hunk 3 above)
Comment (lcspinter) [reply data/conf/iceberg/hive-site.xml:1]: I've discussed this with @deniskuzZ, and he advised that I should open a separate PR to remove this check around locking and update the unit tests which are expecting this exception.

Comment (marton-bod) [reply data/conf/iceberg/hive-site.xml:1]: Sounds good! So that we don't forget, let's create a follow up Jira to remove the new iceberg-test flag once that locking PR has gone in.

File: itests/util/src/main/java/org/apache/hadoop/hive/cli/control/CliConfigs.java [hunk 5]
```diff
@@ -756,4 +756,40 @@ public KuduNegativeCliConfig() {
       }
     }
   }
+
+  public static class IcebergCliConfig extends AbstractCliConfig {
+
+    public IcebergCliConfig() {
+      super(CoreCliDriver.class);
+      try {
+        setQueryDir("iceberg/iceberg-handler/src/test/java/queries/positive");
```
Comment (marton-bod) [inline itests/util/src/main/java/org/apache/hadoop/hive/cli/control/CliConfigs.java:765]: Shouldn't the `.q` files go under `src/test/queries/...` instead of `src/test/java/queries/...`?

Comment (lcspinter) [reply itests/util/src/main/java/org/apache/hadoop/hive/cli/control/CliConfigs.java:765]: You are right, this is a typo

Comment (marton-bod) [reply itests/util/src/main/java/org/apache/hadoop/hive/cli/control/CliConfigs.java:765]: Right, but the q files need to be moved too I think

**Review by marton-bod — APPROVED** [review]

**Comment by lcspinter** [comment]
Thanks, @marton-bod and @pvary for the review.
